### DISCUSSION TITLE
Explain the working of Array and List

### DISCUSSION THREAD

**Discussion by mishelshaji** [mishelshaji/Learn-DotNet#6] [Q&A]
Topics to cover:
👉 Syntax.
👉 Namespace.
👉 Different Methods to add, update, delete and search elements.
👉 Situation to use Array and List.
👉 Memory usage and performance.
👉 Resizing an array.
👉 Different methods to loop over the elements.

**Comment by Jiby-Joseph-123** [comment]
# Array

Arrays are used when there is a need to use many variables of the same type. The array is static, and the length of the array is defined at the time of array declaration.
The Array class is the base class for all the arrays in C#. It is defined in the  System namespace.

 ### Syntax  
``` csharp
type[] arrayName;

int[] newArray = new int[5];
```

The array name is newArray,  int data type, and size is 5.

 ### Declare and set array values:

``` csharp
int[] newArray = new int[]{1,2,3,4};
```
 ### Add

To add an element to an array with an index:
``` csharp
newArray[4]=5;
```
### Update

```csharp
int[] numbers = { 2,4,6,8};
for (int i = 0; i < numbers.Length; i++)
{
    numbers[i] += 2;
}
foreach (int item in numbers)
{
    Console.WriteLine(item);
}
```

Using for loop access each element from the array, add 2 to each item, and update the array.

 ### Remove

Remove the last element from the array:
``` csharp
newArray.pop();
```
remove the first element from the array:
``` csharp
newArray.shift();
```
 ### Get element

To get all elements from an array:
``` csharp
foreach (string i in newArray) 
{
  Console.WriteLine(i);
}
```

For loop for get element from the array:
``` csharp
for(int i =0; i<newArray.Lengh; i++)
{
Console.WriteLine(i);
}
```
 ### Search element

Search an element from the array:
``` csharp
Console.WriteLine("Enter the name to be find");
string key = Console.ReadLine();
Boolean flag = false;
foreach (var item in newArray)
{
    if (key == item)
    {
        flag = true;
        break;
    }
}
if (flag == true)
{
    Console.WriteLine("item found");
}
else
{
    Console.WriteLine("No matches value ");
}
```


### Resizing
```csharp
int[] numbers = { 2,4,6,8};
Console.WriteLine("Length of the array: " numbers.Length); // Length of the array: 4

Array.Resize(ref numbers, 6);
Console.WriteLine("Length of the array: "numbers.Length); // Length of the array: 6
```


# LIST


The list is equivalent to the ArrayList. It is a collection of strongly typed objects. Dynamic memory allocation.

 ### Syntax:
``` csharp
    var lst = new List<int>();
```
 ### Add

To add an element to a list:
 ``` csharp
lst.Add(1);
```
Use the AddRange() method to add all the elements from an array or another collection to the List.
``` csharp
string[] fruits = new string[2]{"Apple","Orange"};


string[] color = new string[2]{"Rose","Orange"};

color.AddRange(fruits);
```
 ### Remove

Remove an element  from a list:
Using value
``` csharp
color.remove("Orange");
```
Using index
``` csharp
color.removeAt(1);
```
 ### Get element

To get all elements from the list.
``` csharp
foreach(var c in color)
{
Console.WriteLine(c);
}
```

### Memory usage and performance

Memory space is reserved for an array and cannot be changed at runtime. If the value to be stored in the array is less than the length of the array, then the remaining space is filled with default values. If we try to input values more than the size of the array it throws an error.

In list memory is allocate dynamically. So we can use space as per our need.

**Reply by mishelshaji** [reply]
Try to cover all points mentioned in the question.

**Comment by vaisak-m** [comment]
**Array** has a pre-defined size; we give an array size while initializing the array. Array is used mostly static, whereas **List** is dynamic. In list we can enter the values without initializing the size first.

### Array

**Syntax**
```csharp

datatype[] arrayName = new datatype[arraySize];
int[] arr = new int[5]; // Array of size 5.
```
**Adding Element**
```csharp
arr[0] = 10;  //Zeroth index is given value 10.
```
**Replacing an Element**
```csharp
arr[0] = 12;
```
**Delete an Element**
```csharp
arr[0] =  ; //Adding a space, we cannot delete an element. We can replace.
```

**Search an element**
```csharp
Console.WriteLine("Enter the element to be find");
string key = Console.ReadLine();
var search;
int count = 0;
foreach(var item in arr)
{
     if (item == search)
      {
         count ++;
       }
 }
if (!count = 0)
{
    Console.WriteLine("item found");
}
else
{
    Console.WriteLine("No matches value ");
}
```
### List
**Syntax**
```csharp
int lst = new List<int>();
```
**Add Elements**
```csharp
lst.Add(1) = 23; //Add at index 1.
```
**Remove Element**
```csharp
lst.Remove(23); //Remove 23 from lst.
lst.RemoveAt(0); //Remove element at index 0.
```
### Memory Usage.
List uses more memory than array, since list are built on top of array. Arrays are efficient in terms of memory usage. For speed array is better because array size is already initialized, and list allocates memory on runtime so it slower compared to array.

**Array Resizing**
```
csharp
Array.Resize(ref arr, arr.Length + 5); //Adds space for 5 more elements.
```
**Methods to loop.**

using for
```csharp
for( i = 0; i < arr.length; i++)
{
Console.WriteLine(i);
}
```

using for each
```csharp
foreach(var item in arr)
{
Console.WriteLine(item);
}

```

**Reply by mishelshaji** [reply]
Try to cover all points mentioned in the question.

**Comment by rashid234** [comment]
**Array in c#**
is a collection of  similar data types with a fixed size.
  
declaration:-
int[] arr = new int[5];

initialization:-
int[] arr= new int[]{1, 2, 3, 4};

for(int i = 0; i<arr.Length;i++)
{
arr[i] = Console.ReadLine();
}

adding elements to array
arr[3] = 2;
removing elements from array
arr[3] = 0;

displaying :-
for(int i = 0; i<arr.Length;i++)
{
 Console.WriteLine(arr[i]);
}

**List**

List in c# is also a type of  collection of data items but it doesnot have a fixed size.

syntax
var lst = new List<int>();

adding elements in list
lst.add(itemToAdd);
deleting elements from list
lst.remove(itemToDelete);
lst.removeAt(index of the itemToDelete);

**Reply by mishelshaji** [reply]
Please try to cover all points mentioned in the question. Also, format the code with markdown.

**Comment by sarath534** [comment]
# Working of Array and List.
## Array:
An array is a fixed-size collection of elements of the same data type. The size of an array is determined when it is created, and cannot be changed afterward. Elements in an array are accessed by their index, which is an integer value that starts at 0.
#### Example:
```csharp
int[] numbers = new int[5];  // create an array of 5 integers

numbers[0] = 1; //assign 1 to the first element of the array
numbers[1] = 2; //assign 2 to the second element of the array

//accessing the element
int firstNumber = numbers[0]; //firstNumber variable will contain 1
```
### Syntax:

 ## Array declaration: 
```csharp
int[] evenNums;  // integer array
string[] cities; // string array
```
## Array Declaration & Initialization
```csharp
int[] evenNums = new int[5]{ 2, 4, 6, 8, 10 }; 
string[] cities = new string[3]{ "Mumbai", "London", "New York" };
```
 ## Short Syntax of Array Declaration
```csharp
int[] evenNums = { 2, 4, 6, 8, 10}; 
string[] cities = { "Mumbai", "London", "New York" 
```
## Accessing Array using for Loop:
Use length property to access elements in the array with conditional statments in for loop:
```csharp
int[] evenNums = { 2, 4, 6, 8, 10 };

for(int i = 0; i < evenNums.Length; i++) // accessing using length property
    Console.WriteLine(evenNums[i]);  

for(int i = 0; i < evenNums.Length; i++)
    evenNums[i] = evenNums[i] + 10;  // update the value of each element
```
## Accessing Array using foreach Loop:
For each element is use to read values of array without index.
```csharp
int[] evenNums = { 2, 4, 6, 8, 10}; 
string[] cities = { "Mumbai", "London", "New York" }; 

foreach(var item in evenNums) //accessing using foreach
    Console.WriteLine(item);   

foreach(var city in cities)
    Console.WriteLine(city);  
```
# Methods to add elements into array
## Array.SetValue() Method
```csharp
array.SetValue(value: 100, index: 0);
array.SetValue(value: 101, index: 1);
array.SetValue(value: 102, index: 2);
```
## Add Values to a C# Array by Using Lists
``` csharp
var list = new List<int>();
list.Add(100);
list.Add(101);
list.Add(102);
var array = list.ToArray();
```
## Method to remove element from array:
```csharp

```
# List:
A list is a dynamic collection of elements in C#. Unlike an array, a list can grow or shrink in size as needed. Elements in a list are also accessed by their index, but a list can also be modified by adding or removing elements.
#### Example:
```csharp
List<int> numbers = new List<int>();  // create a list of integers

numbers.Add(1); //add 1 to the list
numbers.Add(2); //add 2 to the list

numbers.RemoveAt(0); // remove the first element from the list

//accessing the element
int firstNumber = numbers[0]; //firstNumber variable will contain 2
```

**Reply by mishelshaji** [reply]
Please try to cover all points mentioned in the question.

**Comment by sreelakshmiChandran** [comment]
# Array
The array consists of different values and data of the same data types.
- Array initialization
```csharp
int[] arr1= new int[5];              // method 1
var arr2 = new int[5];           // method 2
int[] arr3 = { 2, 8, 9, 6, 4 };  // method 3
int[] arr4 = new int[] { 2, 8, 9, 6, 4 }; // method 4
int[] arr5 = new int[5] { 2, 8, 9, 6, 4 }; // method 5
```
- Accessing elements from an array
```csharp
// method 1
int[] num = {1, 2, 3, 4, 5, 6};
Console.WriteLine(num[2]);
              // output is 3
// method 2
int[] num = {1, 2, 3, 4, 5, 6};

foreach(var item in num)
{
    Console.WriteLine(item);
}

//method 3
for(int = i; i <= num.length; i++;)
{
  Console.WriteLine(num[i]);
}
```
- Update an element in an array
```csharp
int[] num = {1, 2, 3, 4, 5, 6};
num[4] = 9;  // method 1


for(int = i; i <= num.length; i++;)
{
  if(num[i] == 3)
  {
    num[i] = 5;
  }
}                           //method 2

```
- Delete elements from an array
```csharp
int[] num = {1, 2, 3, 4, 5, 6};

Array.Clear(num); // method 1

num.pop(); // method 2

num.shift();  // method 3

```
- Resizing of an array is not possible in c#, but an array can be replaced by another array.

# List
The list is equivalent to the ArrayList. It is a collection of strongly typed objects. Dynamic memory allocation.

- Initializing a list
```csharp
List<int> list = new List<int>();  // method 1
var lst = new List<int>() { 40, 50 };   // method 2
var lst1 = new List<int> { 1, 2, 3, 4 };   // method 3
var lst2 = new List<int> { 4, 5, 6 };   // method 4
var lst2 = new int[] { 4, 5, 6 };   // method 5
```

- Adding elements to the List
```csharp
var lst = new List<int>() { 40, 50 };
lst.Add(1);  // method 1

var lst1 = new List<int> { 1, 2, 3, 4 };
var lst2 = new List<int> { 4, 5, 6 };
lst1.AddRange(lst2);   // method 2
```
- Accessing an element from list
```csharp
var lst = new List<int>() { 40, 50 };
Console.WriteLine(lst[0]);     // method 1
Console.WriteLine(lst.Count);

foreach (var item in lst)       
{
    Console.WriteLine(item);
}                                                   // method 2
```

- Deleting elements from List
```csharp
var lst1 = new List<int> { 1, 2, 3, 4 };
lst1.Remove(2);   // method 1
lst1.RemoveAt(2);   // method 2 
lst1.RemoveRange(0, 2);   // method 3
```
## Array Vs List
- The list is better for frequent insertion and deletion, whereas Arrays are much better suited for frequent access of elements.
- List occupies much more memory as every node defined in the List has its own memory set whereas Arrays are memory-efficient data structures.
- List is mostly used when we want to remove or add data as the array is more expensive.

**Comment by Adarshv12** [comment]
# Array
An array is used to store elements of the same data type. Can not change the size of the array after it is declared.

**Syntax**
```csharp
datatype[] arrayName = new datatype[size]
```

**Set array values**

```csharp
int[] arrayElements = new int[]{1,2,3,4,5};
```

**To add element to an array**

```csharp
arrayName[index] = value; 
```

```csharp
arrayName.push();
```

**Search an element in an array**
```csharp
Console.WriteLine("Enter the name to be find");
string key = Console.ReadLine();
Boolean flag = false;
foreach (var item in newArray)
{
    if (key == item)
    {
        flag = true;
        break;
    }
}
if (flag == true)
{
    Console.WriteLine("item found");
}
else
{
    Console.WriteLine("No matches value ");
}
```

**Remove element from an array**

```csharp
arrayName.pop();
```

```csharp
arrayName.shift();
```

**Array resizing**

```csharp
Array.Resize(ref arr, arr.Length + 5);

```
**Methods to loop array**

```csharp
foreach(var item in arrayName)
{
Console.WriteLine(item);
}

for(i=0,i<arrayName.Length;i++)
{
Console.WriteLine(arrayName[i]);
}
```

# List

List is equivalent to the ArrayList. It is a collection of strongly typed objects. Dynamic memory allocation.

**Syntax:**
```csharp
int listName = new List<int>();
```

**Add elements to a list**
```csharp
listName.add(element);
```
**Remove element from a list**
```csharp
listName.RemovAt(index);

listName.Remove('element');
```

**To display list elements**
```csharp
foreach(var item in listName)
{
Console.WriteLine(item);
}
```
### Memory usage

Since lists are constructed on top of arrays, they utilize more memory than arrays do. The use of RAM is effective for arrays. Because the size of an array is already initialized, arrays are faster than lists in terms of speed. Lists must allocate memory as they go, which makes them slower.

**Comment by riya45810** [comment]
### Array
Array used to store multiple values in a single variable.
**Declaring array**
```csharp
// string array
string[] fruits; 
// integer array
int[] odds; 
```
**Inserting elements**
```csharp
string[] fruits = {"Apple","Orange","Grapes"};
int[] odds = {1,3,5,7,9};
```
**Accessing elements**
```csharp
string[] fruits = {"Apple","Orange","Grapes"};
// outputs Orange
Console.WriteLine("fruits[1]");
```

```csharp
using System;
namespace CsharpProgram
{
class Program
{
 int[] arr;
 Console.WriteLine("Enter a size:");
 int size = int.Parse(Console.ReadLine());
```
 **adding element to array**
```csharp
 Console.WriteLine("Enter elements: ");
  for (int i = 0; i < size; i++)
       {
            arr[i] = Console.ReadLine();
       }
```
**Array resizing**
```csharp
int[] arr = new int[5];
// resize from 5 to 7
Array.Resize(ref arr,7);
```
**searching a particular entered element in array**
```csharp
 Console.WriteLine("Enter a element: ");
 int num = Console.ReadLine();
 int flag = 0;
 for (int i = 0; i < size; i++)
        {
            if (arr[i] == num)
            {
                Console.WriteLine("Number founded! :" + arr[i]);
                flag++;
                break;
            }

        }
 if (flag == 0)
  {
            Console.WriteLine("Number not found!!");
  }
```
**Removing element**
```csharp
// for removing last element from array
arr.pop();
// for removing first element from array
arr.shift();
```
**Loop over elements**
```csharp
string[] fruits = {"Apple","Orange","Grapes"};
// for loop
for(int i=0; i<fruits.Length; i++)[
  Console.WriteLine(fruits[i]);
}
// foreach loop
foreach(var item in fruits){
  Console.WriteLine(item);}
```
**Memory usage**
C# arrays are stored by reserving a block of memory that equals the amount needed to hold the elements
### List
List is the collection of strongly typed objects.
**Declaring List**
```csharp
List<int> oddNumbers = new List<int>();
List cities = new List<string>();
```
**adding elements**
```csharp
oddNumbers.Add(3);
oddNumbers.Add(5);
cities.Add("New York");
cities.Add("London");
```
**Accessing**
```csharp
Console.WriteLine(oddNumbers[0]);
```
**Loop over elements**
```csharp

// for loop
for(int i=0; i<oddNumbers.Count; i++)[
  Console.WriteLine(oddNumbersi]);
}
// foreach loop
oddNumbers.ForEach(num => Console.WriteLine(oddNumbers + ",");
```
**Removing element**
```csharp
// element passed
oddNumbers.Remove(3);
// index passed
oddNumbers.RemoveAt(0);
```
**Memory Usage**
List occupies much more memory as every n ode defined the list has its owm memory set whereas array are memory efficient data structure.

**Comment by thanoofnaina** [comment]
### Array
An array is used to store elements of the same data type. Can not change the size of the array after it is declared. 

**Syntax**
```csharp
datatype[] arrayName = new datatype[size]
```

**Set array values**
```csharp
int[] arrayName = new int[]{a,b,c};
```

**To add element to an array** 
```csharp
arrayName[index] = value;
```

**Loop over elements**
```csharp
string[] fruits = {"Apple","Orange","Grapes"};
// for loop
for(int i=0; i<fruits.Length; i++)[
  Console.WriteLine(fruits[i]);
}
// foreach loop
foreach(var item in fruits){
  Console.WriteLine(item);}
```

**Search an element in an array**
```csharp
Console.WriteLine("Enter a element: ");
int num = Console.ReadLine();
int flag = 0;
for (int i = 0; i < size; i++)
        {
            if (arr[i] == num)
            {
                Console.WriteLine("Number founded! :" + arr[i]);
                flag++;
                break;
            }         }
if (flag == 0)
  {
            Console.WriteLine("Number not found!!");
  }
```

**Removing element**
```csharp
// for removing last element from array
arr.pop();
// for removing first element from array
arr.shift();
```

**Memory usage**
C# arrays are stored by reserving a block of memory that equals the amount needed to hold the elements

### List
List is the collection of strongly typed objects, with a size that can be changed easily.

**Declaring List**
```csharp
List<int> numbers = new List<int>();
```

**adding elements**
```csharp
numbers.Add(1);
numbers.Add(10);
```

**Accessing**
```csharp
Console.WriteLine(numbers[0]);
```

**Loop over elements**
```csharp // for loop
for(int i=0; i<oddNumbers.Count; i++)[
  Console.WriteLine(oddNumbersi]);
}
// foreach loop
oddNumbers.ForEach(num => Console.WriteLine(oddNumbers + ",");
```

**Removing element**
```csharp
// element passed
numbers.Remove(3);
// index passed
numbers.RemoveAt(0);
```

**Memory Usage**
List occupies much more memory as every n code defined the list has its own memory set whereas array are memory efficient data structure.

**Comment by Nandana-04** [comment]
# Array 
An array is a collection of similar data elements stored at contiguous memory locations. It is the simplest data structure where each data element can be accessed directly by only using its index number.
Syntax :
```csharp
dataType[] arrayName = new dataType[size];

eg : 
int[] firstNumbers = new int[6];
string[] secondArray = new string[]{ "Cat" , "Dog" , "Cow" };
```

In an array ,  namespace refers to the scope in which the array variable is defined .

### To add element to an array :-
```csharp
arrayName[Index] = Value;
eg : 
array[2] = 22;

- arrayName.push("Value");

```
### To delete an  element to an array :-
```csharp
 arrayName.pop();
```
```csharp
arrayName.shift();
```
### To update  an  element to an array :-
```csharp
arrayName[Index] = UpdatedValue ;
eg : 
array[1] = 3 ;
```
### To search  an  element to an array :-
```csharp
string[] arr = new string[]{ "A" , "B" , "C" , "D"};
bool flag = false;
            Console.WriteLine("enter the name to search");
            var name = Console.ReadLine();
            for (int i = 0; i < arr.Length; i++)
            {
                if (name == arr[i])
                {
                    Console.WriteLine("name has been found");
                    Console.WriteLine("position of the name");
                    Console.WriteLine(i);
                    flag = true;

                }
            }
            if (flag != true)
            {
                Console.WriteLine("name not found");

            }

```
# List
List is a class that contains multiple objects of the same data type that can be accessed using an index
Syntax : 
```csharp
List<DataType> ListName = new List<DataType> () ;
eg : 
List<int> numbers = new List<int> { 1, 2, 3, 4, 5 } ;
```

- To add element to an list  :-
```csharp

ListName.Add(value) ;
eg :- 
numbers.Add(6) ;
```
- To delete  element to an list  :-
```csharp
ListName.RemoveAt(value) ;
eg :- 
numbers.RemoveAt(1);
```
- To update  element to an list  :-
```csharp
ListName[Index] = newValue ;
eg :- 
numbers[1] = 7 ;
```
### Situation to use Array and List.
 In C# , array and list are used to store collections of items . But , they are used in different situations .
                 Array is a fixed - size  collection of items , which means the size of an array cannot be changed once it is created . Arrays are used when the number of items to be stored is known and will not change , such as when reading data from a file of known length .
                 List is a dynamic array , which means it can be resized as needed . Lists are used when the number of items to be stored is not known in advance , or when items will be frequently added or removed . Lists are more convenient to use than arrays when working with dynamic data .
 
### Memory usage and performance.
An array is a fixed-size data structure , meaning that the size of the array is determined when it is created and cannot be changed afterwards . This can lead to memory efficiencies because the memory for the array can be allocated in a single block . Arrays also have better performance when it comes to random access , because the elements are stored in contiguous memory locations , making it easy to calculate the memory location of any element given its index .

List is a dynamic data structure , meaning that the size of the list can be changed as items are added or removed . This can lead to slightly less memory efficiencies because the memory for the list needs to be reallocated each time the size of the list changes . Lists also have slightly less performant than arrays when it comes to random access , because the elements are stored in non-contiguous memory locations , making it slightly more difficult to calculate the memory location of any element given its index .

### Resizing an array
We cannot resize an array in C#, but using Array.Resize we can replace the array with a new array of different size.
eg :-
```csharp
// Create an object .
char[] ch = new char[10];
ch[0] = 'a';
ch[1] = 'b';
// Resize array .
Array.Resize<char>(ref ch, 10);
// Set value for new elements .
ch[2] = 'c';
ch[3] = 'd';
ch[4] = 'e';
ch[5] = 'f';
ch[6] = 'g';
ch[7] = 'h';
ch[8] = 'i';
ch[9] = 'j';
Console.WriteLine("New Array: " + new string(ch));
```
### Different methods to loop over the elements . 

- For Loop
This is the most basic looping construct in C# . It allows you to iterate over a collection by specifying a starting point , an ending point , and an increment .
```csharp
for (int i = 0; i < arr.Length; i++)
            {
                Console.WriteLine(arr[i].Length);
            }
```
- Foreach Loop
This loop is designed specifically for iterating over collections . It automatically iterates over each element in the collection without the need to specify the starting and ending points .
```csharp
foreach (int value in myArray)
{
    Console.WriteLine(value);
}

```
- While Loop
This loop allows you to iterate over a collection while a certain condition is true .
```csharp
int i = 0;
while (i < myArray.Length)
{
    Console.WriteLine(myArray[i]);
    i++;
}

```
- Do - While Loop
This loop is similar to the while loop , but the body of the loop is executed at least once before the condition is evaluated .
```csharp
int i = 0;
do
{
    Console.WriteLine(myArray[i]);
    i++;
}
while (i < myArray.Length);

```

**Comment by Arjunprasad-22** [comment]
# Array

Arrays are used to store multiple values in a single variable, instead of declaring separate variables for each value. An array is a container object that holds a fixed number of values of a single type and the length of an array is established when the array is created.

## Syntax

### An array can be declared using by specifying the type of its elements with square brackets.

**Example: Array Declaration**
```csharp
int[] evenNums;  // integer array

string[] cities; // string array
```

### The following declares and adds values into an array in a single statement.

**Example: Array Declaration & Initialization**
```csharp
int[] evenNums = new int[5]{ 2, 4, 6, 8, 10 }; 

string[] cities = new string[3]{ "Mumbai", "London", "New York" };
```
Above, evenNums array can store up to five integers. The number 5 in the square brackets new int[5] specifies the size of an array. In the same way, the size of cities array is three. Array elements are added in a comma-separated list inside curly braces { }.

### Arrays type variables can be declared using var without square brackets.

**Example: Array Declaration using var**
```csharp
var evenNums = new int[]{ 2, 4, 6, 8, 10}; 

var cities = new string[]{ "Mumbai", "London", "New York" }; 
```
If you are adding array elements at the time of declaration, then size is optional. The compiler will infer its size based on the number of elements inside curly braces, as shown below.

**Example: Short Syntax of Array Declaration**
```csharp
int[] evenNums = { 2, 4, 6, 8, 10}; 

string[] cities = { "Mumbai", "London", "New York" }
```
### The following example demonstrate invalid array declarations.

**Example: Invalid Array Creation**
```csharp
//must specify the size 
int[] evenNums = new int[]; 

//number of elements must be equal to the specified size 
int[] evenNums = new int[5] { 2, 4 };

//cannot use var with array initializer
var evenNums = { 2, 4, 6, 8, 10}; 
```
### Different array methods

**Example**
```csharp
int[] newArray = new int[]{1,2,3,4};

//To add an element to array with index:
newArray[4]=5;

//Remove last element from the array:
newArray.pop();

//remove first element from the array:
newArray.shift();

//To get all element from an array:
foreach (string item in newArray) 
{
  Console.WriteLine(item);
}

//For loop for get element from array:
for(int i =0; i<newArray.Lengh; i++)
{
Console.WriteLine(i);
}

//Search an element from array:
Console.WriteLine("Enter the name to be find");
string key = Console.ReadLine();
Boolean find = false;
foreach (var item in newArray)
{
    if (key == item)
    {
        find = true;
        break;
    }
}
if (find == true)
{
    Console.WriteLine("item found");
}
else
{
    Console.WriteLine("No matches value ");
}
```
# LIST
List is equivalent of the ArrayList. It is a collection of strongly typed objects. Dynamic memory allocation.

## Syntax:
```csharp
    var list = new List<int>();

//To add element to a list:
list.Add(1);
//Use the AddRange() method to add all the elements from an array or another collection to List.

string[] fruits = new string[2]{"Grapes", "Peaches"};

string[] color = new string[2]{"Jasmine", "Sunflower"};

color.AddRange(fruits);

//Remove an element from a list: Using value
color.remove("Peaches");

//Remove an element from a list: Using index
color.removeAt(1);

//To get all elements from list:
foreach(var item in list)
{
Console.WriteLine(item);
}
```

**Comment by kavitha-BS** [comment]
[array_list.odt](https://github.com/mishelshaji/Learn-DotNet/files/10527977/array_list.odt)

**Comment by sifanaks** [comment]
### Array
Arrays are used to store multiple values in a single variable.
To declare an array, define the variable type with square brackets:
eg:
````c#
string[] cars;
````
Array is defined in the System Namespace.
**Add elements to an Array**
- Array Index Initializer
Using this approach, we can set the value using the specified index or position of an array.
Example:
```` c#
var array = new int[3];
````
- Array.SetValue() Method
The next approach is to use the SetValue() method available in Array class:
````c#
array.SetValue(value: 100, index: 0);
array.SetValue(value: 101, index: 1);
array.SetValue(value: 102, index: 2);
````
We can directly invoke the SetValue() method on an array itself. For that we need to provide the value as the first parameter and the index as the second parameter.
**Deleting elements from an array**
- RemoveAt() Method
- The RemoveAt method in C# is used to remove an element in a list at a position you set.vWe have to convert the array into a list.
To remove an element, set the index from where you want to eliminate the element. The following is to remove element from the 3rd position −
````c#
subjects.RemoveAt(2);
````
**Searching elements in an Array**
- Array.Find()
The Array.Find() method searches for an element that matches the specified conditions using predicate delegate, and returns the first occurrence within the entire Array.
   Syntax:
````c#
public static T Find<T>(T[] array, Predicate<T> match);
````
- Array.FindAll()
The Array.FindAll() method returns all elements that match the specified condition.
Syntax:
````c#
public static T[] FindAll<T>(T[] array, Predicate<T> match)
````
- Array.FindLast()
The Array.Find() method returns the first element that matches the condition. The Array.FindLast() method returns the last element that matches the specified condition in an array.
Syntax:
````c#
public static T FindLast<T>(T[] array, Predicate<T> match)
````
 Arrays are used when there is a need to use many variables of the same type. It can be defined as a sequence of objects which are of the same data type.
 It is used to store a collection of data, and it is more useful to think of an array as a collection of variables of the same type.
Array memory allocated is static and continues.Users needs to ought to confine track of next memory allocation.
You cannot resize an array in C#, but using Array. Resize you can replace the array with a new array of different size.
**Loop Through an Array**
- For loop
You can loop through the array elements with the for loop, and use the Length property to specify how many times the loop should run.
Example
````c#
string[] cars = {"Volvo", "BMW", "Ford", "Mazda"};
for (int i = 0; i < cars.Length; i++) 
{
  Console.WriteLine(cars[i]);
}
````
- The foreach Loop
There is also a foreach loop, which is used exclusively to loop through elements in an array: Syntax
````c#
foreach (type variableName in arrayName) 
{
  // code block to be executed
}
````

###  List
List in C# is a collection of strongly typed objects. These objects can be easily accessed using their respective index. 
Index calling gives the flexibility to sort, search, and modify lists if required.
In simple, List in C# is the generic version of the ArrayList. 
This ArrayList comes under System.Collection.Generic namespace.
Syntax:
````c#
  List<T>
````
The letter T stands for the type of object. 
Syntax to create list of type strings.
````c#
List<string> AuthorList= new List<string>();
`````
**Add items to list**

To add items and elements in a C# list, you need to use the Add() method.


To insert elements at your desired position in a List, you can use the following 
syntax: 
````C#
public void Insert (int index, T item);
````
To add elements at the end, use the method List.Add() method. Method List.AddRange() adds a collection of objects to the end of a given List<T>.
**Remove items from list**:
The Remove() technique will easily eliminate the initial appearance of a particular object from a list. The Remove technique assumes an element as its parameter. 
The RemoveAt technique eliminates an element at the specified zero-based index.
**Find an Element in a C# List**:
The Enumerable.Contains() technique offers an easy and neat means to deduce if a sequence comprises the desired component or not. The alternative solution is to utilize the List.
IndexOf() technique that returns the index of the initial appearance of the desired component in this list and will display -1 in case there is no such component.
**Search  C# List Elements**:
The List.FindAll() technique gives back a list comprising every component that conforms to the given predicate. One can employ it to examine the presence of a component in the provided list. 
**Loop through a list**:
- For each
 The standard option to iterate over the List in C# is using a foreach loop. Then, we can perform any action on each element of the List. The following code example demonstrates its usage by displaying the contents of the list to the console. 
````C#
using System;
using System.Collections.Generic; 
public class Example{  
  public static void Main()
    {      
  List<int> nums = new List<int>() { 1, 2, 3, 4, 5 };  
       foreach (var i in nums)
 {          
  Console.WriteLine(i);    
    } 
   }
}
````
- list<T>.Foreach
Another good alternative to iterate through a list is using List<T>.ForEach(Action<T>) method. It performs the specified action on each element of the list. It accepts a delegate to perform on each element of the List. The following example shows how to print the contents of a list using the Action<T> delegate.
````c#
using System;
using System.Collections.Generic; 
public class Example
{    
public static void Main()  
  {        List<int> nums = new List<int>() { 1, 2, 3, 4, 5 };     
    nums.ForEach(i => Console.WriteLine(i));
    }
````
- For loop
   It is an index based looping mechanism.
````C#
using System;
using System.Collections.Generic;
 public class Example{   
 public static void Main()   
 {      
  List<int> nums = new List<int>() { 1, 2, 3, 4, 5 };    
     for (var i = 0; i < nums.Count; i++) {  
          Console.WriteLine(nums[i]);  
      }  
  }
}
````
